File: F/1865-finding-pairs-with-a-certain-sum/main.go
package main

import "fmt"

type FindSumPairs struct {
	M1 map[int]int
	M2 map[int]int
	N2 []int
}

func Constructor(nums1 []int, nums2 []int) FindSumPairs {
	m1, m2 := map[int]int{}, map[int]int{}
	for _, v := range nums1 {
		m1[v]++
	}
	for _, v := range nums2 {
		m2[v]++
	}
	return FindSumPairs{M1: m1, M2: m2, N2: nums2}
}

func (this *FindSumPairs) Add(index int, val int) {
	o := this.N2[index]
	this.M2[o]--
	this.N2[index] += val
	this.M2[this.N2[index]]++
}

func (this *FindSumPairs) Count(tot int) int {
	ret := 0
	for k, v := range this.M1 {
		v2, ok := this.M2[tot-k]
		if ok {
			ret += v * v2
		}
	}
	return ret
}

func main() {
	fmt.Println()
}

File: F/1865-finding-pairs-with-a-certain-sum/main.rs
use std::collections::HashMap;
struct FindSumPairs {
  m1: HashMap<i32, i32>,
  m2: HashMap<i32, i32>,
  s2: Vec<i32>,
}

impl FindSumPairs {
  fn new(nums1: Vec<i32>, nums2: Vec<i32>) -> Self {
    let mut m1: HashMap<i32, i32> = HashMap::new();
    let mut m2: HashMap<i32, i32> = HashMap::new();

    nums1.iter().for_each(|&v| {
      m1.entry(v).and_modify(|x| *x += 1).or_insert(1);
    });
    nums2.iter().for_each(|&v| {
      m2.entry(v).and_modify(|x| *x += 1).or_insert(1);
    });
    FindSumPairs {
      m1: m1,
      m2: m2,
      s2: nums2,
    }
  }

  fn add(&mut self, index: i32, val: i32) {
    let index = index as usize;
    self.m2.entry(self.s2[index]).and_modify(|x| *x -= 1);
    self.s2[index] += val;
    self
      .m2
      .entry(self.s2[index])
      .and_modify(|x| *x += 1)
      .or_insert(1);
  }

  fn count(&self, tot: i32) -> i32 {
    let mut ans: i32 = 0;
    for (k, v) in &self.m1 {
      let diff = tot - k;
      if self.m2.contains_key(&diff) {
        ans += self.m1.get(&k).unwrap() * self.m2.get(&diff).unwrap();
      }
    }

    ans
  }
}
